File: site/testsuites/testsuite-javascript/__tests__/tbdex/pfi/creatingQuotes.test.js
import { Rfq, Quote, Parser } from '@tbdex/http-server';
import { DevTools } from '@tbdex/http-client';
import { DidDht } from '@web5/dids';
import { OfferingsApiProvider } from './offeringsApiProvider'
import { ExchangesApiProvider } from './exchangesApiProvider'
import { MockDataProvider } from '../../utils/mockDataProvider'
import { vi, test, expect, describe, beforeAll } from 'vitest';

let pfiDid;
let senderDid;
let message;
let dataProvider = new MockDataProvider();
let offeringsApiProvider;
let exchangesApiProvider;
let mockOffering;

describe('PFI: Quotes', () => {
  beforeAll(async () => {
    // Set up providers and DID
    pfiDid = await DidDht.create({
      options:{
        publish: true,
        services: [{
            id: 'pfi',
            type: 'PFI',
            serviceEndpoint: 'https://example.com/'
        }]
      }
    });

    senderDid = await DidDht.create({
      options: { publish: true }
    })
    
    offeringsApiProvider = new OfferingsApiProvider(pfiDid);
    exchangesApiProvider = new ExchangesApiProvider();

    // Configure Mocks

    message = await DevTools.createRfq({
      sender: senderDid,
      receiver: pfiDid
    });
    await message.sign(senderDid);
 
    mockOffering = DevTools.createOffering({
      from: pfiDid.uri,
      offeringData: DevTools.createOfferingData()
    })
    await mockOffering.sign(pfiDid);


    message.offeringId = mockOffering.id;
    offeringsApiProvider.setOffering(mockOffering);


    dataProvider.setupInsert("exchange", "", () => { return });
  });
  
  test('PFI creates offering', async () => {
    // :snippet-start: pfiWriteOfferingJs
    // Write the message to your exchanges database
    await dataProvider.insert('exchange', {
      exchangeid: message.exchangeId,
      messagekind: message.kind,
      messageid: message.id,
      subject: message.subject,
      message: await Parser.parseMessage(message)
    });
  
    //highlight-start
    const offering = await offeringsApiProvider.getOffering(message.offeringId);
    //highlight-end
    // :snippet-end:
  })

  test('PFI creates and signs quote', async () => {
    const offering = mockOffering

    // :snippet-start: pfiCreateQuoteJs
    // Set the Quote's expiration date for 1 day from now
    var quoteExpiration = new Date();
    quoteExpiration.setDate(quoteExpiration.getDate() + 1);

    const quote = Quote.create({
      metadata: {
        from: pfiDid.uri,
        to: message.from,
        exchangeId: message.exchangeId,
        protocol: '1.0'
      },
      data: {
        expiresAt: quoteExpiration.toLocaleDateString('en-us'),
        payin: {
          currencyCode: offering.data.payin.currencyCode,
          amount: '0.01',
          fee: '0.0001',
          paymentInstruction : {
            link: 'https://example.com/paymentInstructions',
            instruction: 'Detailed payment instructions'
          }
        },
        payout: {
          currencyCode: offering.data.payout.currencyCode,
          amount: '1000.00',
          paymentInstruction : {
            link: 'https://example.com/paymentInstructions',
            instruction: 'Detailed payout instructions'
          }
        }
      }
    });
    // :snippet-end:
  
    exchangesApiProvider.setWrite();
  
    // :snippet-start: pfiSignQuoteJs
    await quote.sign(pfiDid);
    dataProvider.insert(quote);
    // :snippet-end:

    const signature = await quote.verifySignature();
    expect(signature).toBeDefined();
  })

})

File: site-new/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/pfi/CreatingQuotesTest.kt
package website.tbd.developer.site.docs.tbdex.pfi

import tbdex.sdk.protocol.models.Rfq
import tbdex.sdk.protocol.models.Quote
import tbdex.sdk.protocol.models.*
import web5.sdk.dids.methods.dht.DidDht
import web5.sdk.crypto.InMemoryKeyManager
import web5.sdk.dids.did.BearerDid
import java.time.OffsetDateTime
import website.tbd.developer.site.docs.utils.*
import web5.sdk.dids.didcore.Service
import web5.sdk.dids.methods.dht.CreateDidDhtOptions
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.*
import java.time.OffsetDateTime.*

class CreatingQuotesTest {

    private lateinit var offeringsApiProvider: OfferingsApiProvider
    private lateinit var exchangesApiProvider: ExchangesApiProvider
    private lateinit var dataProvider: MockDataProvider
    private lateinit var pfiDid: BearerDid
    private lateinit var message: Message

    @BeforeEach
    fun setup() {
        offeringsApiProvider = OfferingsApiProvider()
        exchangesApiProvider = ExchangesApiProvider()
        dataProvider = MockDataProvider()

        val serviceToAdd = Service.Builder()
            .id("pfi")
            .type("PFI")
            .serviceEndpoint(listOf("https://example.com/"))
            .build()

        val options = CreateDidDhtOptions(
            publish = true,
            services = listOf(serviceToAdd),
        )

        pfiDid = DidDht.create(InMemoryKeyManager(), options)

        message = TestData.getRfq(to = pfiDid.uri)
    }

    @Test
    fun `PFI creates and signs quote`() {
        dataProvider.setupInsert("exchange", "") { arrayOf<Any>() }
        offeringsApiProvider.setOffering(message.metadata.id, pfiDid)

        // :snippet-start: pfiWriteOfferingKt
        // Write the message to your exchanges database
        val data = mapOf(
            "exchangeid" to message.metadata.exchangeId,
            "messagekind" to message.metadata.kind,
            "messageid" to message.metadata.id,
            "subject" to message.metadata.from,
            "message" to message.data
        )

        dataProvider.insert("exchange", data)
        //highlight-start
        val offering = offeringsApiProvider.getOffering(message.metadata.id)
        //highlight-end
        // :snippet-end:

        // :snippet-start: pfiCreateQuoteKt
        val quote = Quote.create(
            to = message.metadata.from,
            from = pfiDid.uri,
            exchangeId = message.metadata.exchangeId,
            protocol = "1.0",
            quoteData = QuoteData(
                expiresAt = now().plusDays(10),
                payin = QuoteDetails(
                    currencyCode = offering.data.payin.currencyCode,
                    amount = "250",
                    fee = "0.001",
                    paymentInstruction = PaymentInstruction(
                        link = "https://example.com/paymentInstructions",
                        instruction = "Detailed payout instructions"
                    )
                ),
                payout = QuoteDetails(
                    currencyCode = offering.data.payout.currencyCode,
                    amount = "1248.22",
                    paymentInstruction = PaymentInstruction(
                        link = "https://example.com/paymentInstructions",
                        instruction = "Detailed payout instructions"
                    )
                )
            )
        )
        // :snippet-end:

        exchangesApiProvider.setWrite()

        // :snippet-start: pfiSignQuoteKt
        quote.sign(pfiDid)
        exchangesApiProvider.write(quote)
        // :snippet-end:

        assertNotNull(quote.verify(), "Quote signature is invalid")
    }
}
